aioshelly/rpc_device/wsrpc.py

```diff
@@ -323,41 +327,52 @@ def handle_frame(self, frame: dict[str, Any]) -> None:
     async def _rx_msgs(self) -> None:
         assert self._client
 
-        while not self._client.closed:
-            try:
-                msg = await self._client.receive()
-                self._last_time = time.time()
-                if msg.type == WSMsgType.PONG:
-                    self._schedule_heartbeat()
-                    continue
-                if msg.type == WSMsgType.PING:
-                    await self._client.pong(msg.data)
-                    continue
-                frame = _receive_json_or_raise(msg)
-                _LOGGER.debug("recv(%s): %s", self._ip_address, frame)
-            except InvalidMessage as err:
-                _LOGGER.error("Invalid Message from host %s: %s", self._ip_address, err)
-            except ConnectionClosed:
-                break
-            except Exception:  # pylint: disable=broad-except
-                _LOGGER.exception("Unexpected error while receiving message")
+        try:
+            while not self._client.closed:
+                try:
+                    msg = await self._client.receive()
+                    self._last_time = time.time()
+                    if msg.type == WSMsgType.PONG:
+                        self._schedule_heartbeat()
+                        continue
+                    if msg.type == WSMsgType.PING:
+                        await self._client.pong(msg.data)
+                        continue
+                    frame = _receive_json_or_raise(msg)
+                    _LOGGER.debug("recv(%s): %s", self._ip_address, frame)
+                except InvalidMessage as err:
+                    _LOGGER.error(
+                        "Invalid Message from host %s: %s", self._ip_address, err
+                    )
+                except ConnectionClosed:
+                    break
+                except Exception:  # pylint: disable=broad-except
+                    _LOGGER.exception("Unexpected error while receiving message")
+
+                if not self._client.closed:
+                    self.handle_frame(frame)
+        finally:
+            _LOGGER.debug(
+                "Websocket client connection from %s closed", self._ip_address
+            )
+            self._cancel_heatbeat_and_pong_response_cb()
+            # Ensure the underlying transport is closed
+            # Remove this after aiohttp 3.9.2 is released
+            # as it leaks before this version (and has for
+            # a long time)
+            with contextlib.suppress(Exception):
+                self._client._response.close()  # pylint: disable=protected-access
```

Comment:
This is not the issue, but it is a problem since aiohttp < 3.9.2 won't close the transport on failure